File: probs/2533.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**8)

input = open('input.txt', 'r').readline

def solve(cur, pre, early):
    if dp[cur][early] < N+1:
        return dp[cur][early]
    dp[cur][early] = early

    for nxt in G[cur]:
        if nxt == pre:
            continue

        if early:
            dp[cur][early] += min(
                solve(nxt, cur, 0),
                solve(nxt, cur, 1),
            )
        else:
            dp[cur][early] += solve(nxt, cur, 1)

    return dp[cur][early]

N = int(input())
G = [[] for _ in range(N+1)]
for _ in range(N-1):
    u, v = map(int, input().split())
    G[u].append(v)
    G[v].append(u)

dp = [[N+1]*2 for _ in range(N+1)]
solve(1, 0, 0)
solve(1, 0, 1)
print(min(dp[1]))

File: probs/2533.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX = 1e6,
          INF = 1e9;

int dp[MAX+1][2];
vector<int> G[MAX+1];

int solve(int u, int p, int e)
{
    if (dp[u][e] == INF) {
        dp[u][e] = e;
        for (int v : G[u]) {
            if (v == p) continue;
            dp[u][e] += min(
                e ? solve(v, u, 0) : INF,
                solve(v, u, 1)
            );
        }
    }

    return dp[u][e];
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int N; cin >> N;
    fill(&dp[0][0], &dp[N+1][0], INF);
    while (--N) {
        int u, v; cin >> u >> v;
        G[u].push_back(v);
        G[v].push_back(u);
    }

    solve(1, 0, 0);
    solve(1, 0, 1);
    cout << min(dp[1][0], dp[1][1]) << '\n';

    return 0;
}
